src/field_5x52_int128_impl.h

```diff
@@ -12,268 +6,241 @@
-    VERIFY_BITS_128(&d, 114);
-    /* [d t4+(tx<<48) t3 0 0 c] = [p8 0 0 p5 p4 p3 0 0 p0] */
-    u0 = secp256k1_u128_to_u64(&d) & M; secp256k1_u128_rshift(&d, 52);
-    VERIFY_BITS(u0, 52);
-    VERIFY_BITS_128(&d, 62);
-    /* [d u0 t4+(tx<<48) t3 0 0 c] = [p8 0 0 p5 p4 p3 0 0 p0] */
-    /* [d 0 t4+(tx<<48)+(u0<<52) t3 0 0 c] = [p8 0 0 p5 p4 p3 0 0 p0] */
-    u0 = (u0 << 4) | tx;
-    VERIFY_BITS(u0, 56);
-    /* [d 0 t4+(u0<<48) t3 0 0 c] = [p8 0 0 p5 p4 p3 0 0 p0] */
-    secp256k1_u128_accum_mul(&c, u0, R >> 4);
-    VERIFY_BITS_128(&c, 113);
-    /* [d 0 t4 t3 0 0 c] = [p8 0 0 p5 p4 p3 0 0 p0] */
-    r[0] = secp256k1_u128_to_u64(&c) & M; secp256k1_u128_rshift(&c, 52);
-    VERIFY_BITS(r[0], 52);
-    VERIFY_BITS_128(&c, 61);
-    /* [d 0 t4 t3 0 c r0] = [p8 0 0 p5 p4 p3 0 0 p0] */
+#define secp256k1_fe_sqr_inner fiat_secp256k1_dettman_square
+#define secp256k1_fe_mul_inner fiat_secp256k1_dettman_mul
 
-    a0 *= 2;
-    secp256k1_u128_accum_mul(&c, a0, a1);
-    VERIFY_BITS_128(&c, 114);
-    /* [d 0 t4 t3 0 c r0] = [p8 0 0 p5 p4 p3 0 p1 p0] */
-    secp256k1_u128_accum_mul(&d, a2, a4);
-    secp256k1_u128_accum_mul(&d, a3, a3);
-    VERIFY_BITS_128(&d, 114);
-    /* [d 0 t4 t3 0 c r0] = [p8 0 p6 p5 p4 p3 0 p1 p0] */
-    secp256k1_u128_accum_mul(&c, secp256k1_u128_to_u64(&d) & M, R); secp256k1_u128_rshift(&d, 52);
-    VERIFY_BITS_128(&c, 115);
-    VERIFY_BITS_128(&d, 62);
-    /* [d 0 0 t4 t3 0 c r0] = [p8 0 p6 p5 p4 p3 0 p1 p0] */
-    r[1] = secp256k1_u128_to_u64(&c) & M; secp256k1_u128_rshift(&c, 52);
-    VERIFY_BITS(r[1], 52);
-    VERIFY_BITS_128(&c, 63);
-    /* [d 0 0 t4 t3 c r1 r0] = [p8 0 p6 p5 p4 p3 0 p1 p0] */
+#if (-1 & 3) != 3
+#error "This code only works on a two's complement system"
+#endif
 
-    secp256k1_u128_accum_mul(&c, a0, a2);
-    secp256k1_u128_accum_mul(&c, a1, a1);
-    VERIFY_BITS_128(&c, 114);
-    /* [d 0 0 t4 t3 c r1 r0] = [p8 0 p6 p5 p4 p3 p2 p1 p0] */
-    secp256k1_u128_accum_mul(&d, a3, a4);
-    VERIFY_BITS_128(&d, 114);
-    /* [d 0 0 t4 t3 c r1 r0] = [p8 p7 p6 p5 p4 p3 p2 p1 p0] */
-    secp256k1_u128_accum_mul(&c, R, secp256k1_u128_to_u64(&d)); secp256k1_u128_rshift(&d, 64);
-    VERIFY_BITS_128(&c, 115);
-    VERIFY_BITS_128(&d, 50);
-    /* [(d<<12) 0 0 0 t4 t3 c r1 r0] = [p8 p7 p6 p5 p4 p3 p2 p1 p0] */
-    r[2] = secp256k1_u128_to_u64(&c) & M; secp256k1_u128_rshift(&c, 52);
-    VERIFY_BITS(r[2], 52);
-    VERIFY_BITS_128(&c, 63);
-    /* [(d<<12) 0 0 0 t4 t3+c r2 r1 r0] = [p8 p7 p6 p5 p4 p3 p2 p1 p0] */
+/*
+ * The function fiat_secp256k1_dettman_mul multiplies two field elements.
+ *
+ * Postconditions:
+ *   eval out1 mod 115792089237316195423570985008687907853269984665640564039457584007908834671663 = (eval arg1 * eval arg2) mod 115792089237316195423570985008687907853269984665640564039457584007908834671663
+ *
+ * Input Bounds:
+ *   arg1: [[0x0 ~> 0x1ffffffffffffe], [0x0 ~> 0x1ffffffffffffe], [0x0 ~> 0x1ffffffffffffe], [0x0 ~> 0x1ffffffffffffe], [0x0 ~> 0x1fffffffffffe]]
```

Comment:
Would you willing to format your printed values in the comments with leading zeros to make everything 64 bits (or 32-bits) wide?

Reply:
> ایا مایل هستید مقادیر چاپ شده خود را در نظرات با صفرهای پیشرو فرمت کنید تا همه چیز 64 بیت (یا 32 بیت) عرض داشته باشد؟

Reply:
@andres-erbsen, would you want to do that on fiat's side or should I just manually edit that?

Reply:
I don't have specific preferences, but perhaps a more general solution  [here](https://github.com/mit-plv/fiat-crypto/blob/60efd195b71a696a9f36f5f001d0a92eba96e4a6/src/Stringification/Language.v#L1566) is more likely to be typo-free. I'd be happy to implement this if that's what this PR needs to move forward.

Reply:
I assume it is not too tricky to do it on the fiat side. Let's do that and let the maintainers comment, what else is needed to get this merged :)

Reply:
I believe it is still unclear whether it is better to merge this or simply prove the existing C code correct in VST.  Unfortunately, I don't think any conclusions were reached in the last meeting.

I've offered to move up my priority on creating a VST proof of the correctness of these functions (right after I finish up with the correctness of `secp256k1_modinv64_divsteps_62_var`) so we have something to compare with, but I only work 1 day a week on VST.

Reply:
kindly bumping this as it has been a couple of weeks now. Has there been any decision made yet to (not) integrate that?

Reply:
I've begun work on a VST proof of correctness of the current implementation of `secp256k1_fe_mul_inner`.  I've gone as far as proving the correctness of a functional implementation of the dettman algorithm.  Next up is proving that if the input has suitably bounded inputs then the outputs will be suitably bounded, and then proving the correctness of the actual C implementation by showing none of the operations overflow and none of the VERIFY_CHECK statements fail (given the magnitude preconditions are satisfied).

Speaking only for myself, my thinking is that I complete this proof and then we can look at were we are with regards to these two (unfortunately incompatible) approaches to correctness and decide from there.  Or I give up on my work.  Or you folks call me out in not making progress.

Again, I only work 1 day a week on this project, so I'm guessing several more weeks will still be needed on my end.